### ISSUE TITLE
[PSR-7] ServerRequestInterface: Session missing?

### ISSUE THREAD

**Issue by janwalther** [php-fig/fig-standards#411]
As ServerRequestInterface is supposed to be used as an internal, unmutable representation of the HTTP request I wonder how e.g. a controller could access a session. Of course I understand that sessions are not part of the request (if then only the session-id which gets converted to $_SESSION by PHP). 
But I thought ServerRequestInterface is intended to be used as an encapsulation of all request parameters - and sessions belong to them, don't they?

**Comment by weierophinney** [comment]
I've debated this, but ultimately think this is out-of-scope.

The reason is that sessions are really, really complex, and you cannot just wrap usage in an object easily:

- You don't need sessions on every request. In fact, you typically want to strive to have them enabled as little as possible, as they make client-side caching impossible, and provide significant performance overhead.
- You end up needing to delegate to the `$_SESSION` superglobal regardless. The other superglobals do not have this requirement, as they are typically read-only, not read-write, in usage.
- In order for sessions to work, there's a lot of plumbing you need to do: session configuration, `session_start()`, etc. that will lead to an extensive API to accommodate.
- Sessions are a testing nightmare.

Essentially, adding this to the spec would become an implementation nightmare. And there are already a number of good libraries that wrap only sessions (Aura.Session is excellent, and I've used it in an application that combines PSR-7 with some session functionality).

**Comment by weierophinney** [comment]
@pmjones This can be closed.

**Comment by mindplay-dk** [comment]
Am I missing something really obvious, or why is interoperability on every aspect of HTTP important *except* for sessions? Seems like a pretty big gap.

I understand that the session library in PHP is laden with issues - far from limited to the ones you summarized above. Does that mean we can simply ignore the session issue entirely? Seems like a pretty central thing to most web projects.

Interoperability as far as getting and setting session variables could at least be streamlined, very simply, with an interface like [this one](https://github.com/Wandu/Session/blob/master/src/SessionInterface.php) - why do you need to deal with session semantics at all?

Yes, the semantics of the `session_*()` functions are completely muddled and wonky - but no one is forcing you to implement your session providers on top of that. (actually, having seen enough implementations of that, personally, I'd say that's a pretty lousy idea.)

In fact, if you want to remain interoperable with PSR-7, the standard session library is pretty much out of the question - since one of the key responsibilities of the response abstraction is to abstract the headers that are part of a response, this basically clashes head-on with the session library, which will just send headers willy-nilly, if and when it feels like it. Of course you could work around that, but it would be much simpler and cleaner to just implement custom session handling (as in the library referenced above) rather than fighting the crappy semantics of the standard session library.

Either way, whether somebody prefers to implement a custom session storage facility, or wrap up the standard facility, access to either can be standardized just fine using a simple get/set interface - just simply ignore all the problems that come with the standard session implementations and let implementers choose whether they want that pain or not.

Why is that not an option?

When I realized sessions were left out of this PSR, I felt like I just fell into a black hole ;-)

**Comment by weierophinney** [comment]
Sessions *are* covered, tangentially, via headers. This means you can get the session cookie, and then use it to match to the session store.

Persistence, however, is out of scope for the messages. If you look at every other item in the `ServerRequestInterface`, including cookies, the values are present *in* *the* *message*. This is not the case with sessions, as the actual values are *somewhere* *else*, and that somewhere else can vary widely depending on whether you're using PHP's native `ext/session` or if you're rolling your own session persistence (which we're actually looking at for the next incarnation of ZF, and which I've seen many companies who do massive scaling do).

As such, they *are* something completely separate. I would envision libraries that are able to get the session cookie from the incoming request, and regenerate one to return to the client; these would *consume* PSR-7 in that regard. PSR-7 is a data source for the identifier for both the incoming request and the response sent back to the client, no more, no less.

It also means that you can opt-in to sessions, instead of having them present by default — which you definitely want to do as sessions add a fair amount of overhead that it's best to avoid unless necessary. Use a library such as [aura.session](https://github.com/auraphp/Aura.Session), and *inject* it in controllers that need it. I can tell you already that it works perfectly with PSR-7 (I'm using that combination on my own site currently; aura.session allows injecting the cookie values!).

**Comment by mindplay-dk** [comment]
I understand all of that - still, the problem remains, aura session exists in it's own little void, as does every other implementation, and every implementation is going to have different semantics and different means of interacting with the request/response abstractions; or not interacting with them at all, if you're wrapping the standard session library.

From the consumer's perspective, e.g. the controller, all it should need to care about is getting and setting values - everything else is an implementation detail, most likely handled by your dispatcher, router, framework, project configuration, whatever. What your controllers care about, is having a collection of mutable state associated with the current user's session.

Aura.Session introduces a bunch of other, more high-level concepts, such as flash values, CSRF tokens, etc. - all of which depend on session variables, for which they invent [their own interface](https://github.com/auraphp/Aura.Session/blob/2.x/src/SegmentInterface.php), which is not interoperable with e.g. the one I mentioned above. If the get and set methods had the same signature in those libraries, they could be made interoperable, and things like flash values and CSRF tokens could be factored into components that depend only on a getter/setter interface, making that functionality available regardless of which session library you have underlying.

We have every superglobal neatly abstracted, save for `$_SESSION`. It seems like low-hanging fruit, it seems to important to ignore, and it really, really seems incomplete without it...

**Comment by pmjones** [comment]
(Full disclosure: I'm the lead on Aura.Session.)

I have to back @weierophinney on this one. Sessions are only indirectly and distantly related to HTTP message descriptions. They're not a core functionality of HTTP requests or responses. It may be wise for to begin a Session-related PSR, but session work doesn't belong in an HTTP-related PSR.

**Comment by weierophinney** [comment]
> It seems like low-hanging fruit, it seems to important to ignore, and it really, really seems incomplete without it...

We actually discussed this at some length when developing `ServerRequestInterface`, and felt, for the reasons stated above, that it was not a good fit.

While it is *definitely* painful that the various concrete libraries that exist have different APIs, that does not mean that the functionality belongs in PSR-7. It does suggest that either another PSR is required for sessions — or perhaps the much-discussed-but-never-yet-proposed key/value store, which could certainly be used as a basis for a session store as well.

**Comment by mindplay-dk** [comment]
I get what you're saying, completely.

I don't even disagree as such.

I just don't believe there's a single PHP developer out there who thinks of `$_SESSION` as being a distant relative of `$_GET`, `$_POST`, `$_FILES`, etc. - all the superglobals combined describe the context of the current request. In that sense, `ServerRequest` is incomplete without it - it may fully describe an HTTP request, but it doesn't fully describe a PHP server request.

`ServerRequestInterface` includes other things that aren't related to HTTP at all anyhow, such as the server environment variables - those are part of the request context, just like session variables. The main difference is that server variables aren't surrounded by the wacky semantics and problems that shrouds session variable management - and it's immutable, so it fits with your abstraction, so it was an easy pick. Session variables not so much... They're a hot potato, believe me, I get that ;-)

**Comment by weierophinney** [comment]
> I just don't believe there's a single PHP developer out there who thinks of $_SESSION as being a distant relative of $_GET, $_POST, $_FILES, etc.

Clearly there are, as those developers contributing to the spec and discussions around the spec felt it differed enough to fall outside the scope of the spec. Be careful about generalizations, please.

> ServerRequestInterface includes other things that aren't related to HTTP at all anyhow, such as the server environment variables - those are part of the request context, just like session variables.

I'd argue that the session is **not** part of the request context. It is persistent data *identified* by tokens in the request context, but not *part* of it itself. As such, it's more similar to pulling data from the database.

You cannot compare `$_SERVER` with `$_SESSION` at all, really. `$_SERVER` is data present in the request environment as provided by the web server and/or SAPI, *always*. `$_SESSION` is contingent on a *persistence* *engine*, and sessions may *or* *may* *not* be enabled on any given request. PHP happens to provide a persistence engine via `ext/session`, but it's not the *sole* mechanism possible for providing sessions; even when using it, the mechanism for persisting sessions is mutable (via session wrappers). As I've noted elsewhere, I've seen plenty of systems that eschew PHP's mechanism entirely, and implementing their own session mechanism. On top of that, I've also noted elsewhere that providing access to session variables out-of-the-box involves a ton of behavior that needs to be configurable (to allow configuring how sessions work), and which needs to be opt-in (i.e., *not* enabled by default). This falls completely out-of-scope for the specification, and definitely begs for a specialized solution.

**Comment by tuupola** [comment]
HTTP is stateless which by definition leaves the sessions out of scope. However maybe sessions would be good candidate for new PSR?

**Comment by mindplay-dk** [comment]
Alright, I'll buy all of that.

I really was hoping this PSR would be the *end* of superglobals, but I guess, yeah, sessions call for a separate PSR. That'll probably be another year or so of heated debates, heh ;-)

**Comment by weierophinney** [comment]
@mindplay-dk Keep an eye on the cache discussions. I honestly think the alternate proposal of the key/value store (which I'm hoping will get proposed soon) would be a good fit for this; you'd create a "session-aware" instance and pass that around, allowing you to leverage interfaces, as well as swap out the persistence if you needed to later.
